[clang] Diagnose problematic diagnostic messages

AaronBallman opened the pull request:
Clang has some unwritten rules about diagnostic wording regarding things like punctuation and capitalization. This patch documents those rules and adds some tablegen support for checking diagnostics follow the rules.

Specifically: tablegen now checks that a diagnostic does not start with a capital letter or end with punctuation, except for the usual exceptions like proper nouns or ending with a question.

Now that the code base is clean of such issues, the diagnostics are emitted as an error rather than a warning to ensure that failure to follow these rules is either addressed by an author, or a new exception is added to the checking logic.

AaronBallman committed:
Diagnose problematic diagnostic messages

Clang has some unwritten rules about diagnostic wording regarding
things like punctuation and capitalization. This patch documents those
rules and adds some tablegen support for checking diagnostics follow
the rules.

Specifically: tablegen now checks that a diagnostic does not start with
a capital letter or end with punctuation, except for the usual
exceptions like proper nouns or ending with a question.

Endilll commented on clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:
```diff
@@ -1213,6 +1213,194 @@ static bool isRemark(const Record &Diag) {
+      .Case("Microsoft", true)
+      .Case("Neon", true)
+      .StartsWith("NSInvocation", true) // NSInvocation, NSInvocation's
+      .Case("Objective", true)          // Objective-C, Objective-C++
```
Probably worth a comment that `-` is considered a word boundary, so both Objective-C and Objective-C++ are covered.

AaronBallman replied:
Okay, I did that and dropped the "Objective-C++" so it still fits inline.

erichkeane commented on clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:
```diff
@@ -1213,6 +1213,194 @@ static bool isRemark(const Record &Diag) {
+      return;
+
+    StringRef SelectText = FullDiagText.substr(
+        /*%select{*/ 8, Iter - FullDiagText.begin() - /*%select{*/ 8);
```
instead of magic numbers, can we use 'sizeof' for %select{?

Endilll replied:
You beat me to this! Yeah, that'd be nice.

AaronBallman replied:
I had originally done that but that still needs magic numbers because `sizeof("%select{")` is 9, not 8 (it includes the null terminator). But because the same suggestion came up twice in addition to my initial attempt, I made the changes and used `constexpr` variables to hide the subtraction.

erichkeane reviewed (approved):
1 suggestion on improving the diagnostics, else, LGTM.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-diag-tablegen-checking

AaronBallman committed:
Update based on review feedback

* Improves a comment
* Uses sizeof instead of magic numbers

erichkeane commented on clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:
```diff
@@ -1302,7 +1302,8 @@ static void verifyDiagnosticWording(const Record &Diag) {
 
     size_t BraceCount = 1;
-    auto Iter = FullDiagText.begin() + /*%select{*/ 8;
+    constexpr size_t PercentSelectBraceLen = sizeof("%select{") - 1;
```
Even better :)

erichkeane commented on clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:
```diff
@@ -1355,7 +1357,7 @@ static void verifyDiagnosticWording(const Record &Diag) {
     // Scan backwards to find the opening curly brace.
     size_t BraceCount = 1;
-    auto Iter = FullDiagText.end() - /*}0*/ 3;
+    auto Iter = FullDiagText.end() - sizeof("}0");
```
Does this not need a -1? Only 2 chars, but was/ends up being 3? Or is this offset because 'end' is 1 past end?

AaronBallman replied:
It's because we want to get to the iterator just *before* the `}0`, so -2 gets us to `}` while -3 gets us to the character before the `}`.

Endilll commented on clang/utils/TableGen/ClangDiagnosticsEmitter.cpp:
```diff
@@ -1213,6 +1213,197 @@ static bool isRemark(const Record &Diag) {
+      .Case("Microsoft", true)
+      .Case("Neon", true)
+      .StartsWith("NSInvocation", true) // NSInvocation, NSInvocation's
+      .Case("Objective", true)          // Objective-C (- is a word boundary)
```
```suggestion
      .Case("Objective", true)          // Objective-C (hyphen is a word boundary)
```

AaronBallman committed:
Gracefully handle empty text at the start

This handles cases like '%select{|okay}0 Now' where we're checking for
improper capitalization at the start but have an empty string to check.

AaronBallman committed:
Merge remote-tracking branch 'origin/main' into aballman-diag-tablegen-checking

AaronBallman committed:
Update comment; NFC

AaronBallman committed:
Fix formatting; NFC
